pkg/expr/mathexp/reduce.go

```diff
@@ -80,33 +83,106 @@ func Last(fv *Float64Field) *float64 {
 	return &f
 }
 
+func GetReduceFunc(rFunc string) (ReducerFunc, error) {
+	switch strings.ToLower(rFunc) {
+	case "sum":
+		return Sum, nil
+	case "mean":
+		return Avg, nil
+	case "min":
+		return Min, nil
+	case "max":
+		return Max, nil
+	case "count":
+		return Count, nil
+	case "last":
+		return Last, nil
+	default:
+		return nil, fmt.Errorf("reduction %v not implemented", rFunc)
+	}
+}
+
 // Reduce turns the Series into a Number based on the given reduction function
-func (s Series) Reduce(refID, rFunc string) (Number, error) {
+// if ValueMapper is defined it applies it to the provided series and performs reduction of the resulting series.
+// Otherwise, the reduction operation is done against the original series.
+func (s Series) Reduce(refID, rFunc string, mapper ValueMapper) (Number, error) {
 	var l data.Labels
 	if s.GetLabels() != nil {
 		l = s.GetLabels().Copy()
 	}
 	number := NewNumber(refID, l)
 	var f *float64
-	fVec := s.Frame.Fields[seriesTypeValIdx]
+	series := s
+	if mapper != nil {
+		series = mapper.MapSeries(s)
+	}
+	fVec := series.Frame.Fields[seriesTypeValIdx]
 	floatField := Float64Field(*fVec)
-	switch rFunc {
-	case "sum":
-		f = Sum(&floatField)
-	case "mean":
-		f = Avg(&floatField)
-	case "min":
-		f = Min(&floatField)
-	case "max":
-		f = Max(&floatField)
-	case "count":
-		f = Count(&floatField)
-	case "last":
-		f = Last(&floatField)
-	default:
-		return number, fmt.Errorf("reduction %v not implemented", rFunc)
+	reduceFunc, err := GetReduceFunc(rFunc)
+	if err != nil {
+		return number, err
+	}
+	f = reduceFunc(&floatField)
+	if f != nil && mapper != nil {
+		f = mapper.MapResult(f)
 	}
 	number.SetValue(f)
-
 	return number, nil
 }
+
+type ValueMapper interface {
+	MapSeries(s Series) Series
+	MapResult(v *float64) *float64
+}
```

Comment:
Seems like a good pattern generally. Perhaps:

```
type ReduceMapper interface {
    InputMapper(x *float64) *float64
    OutputMapper(x *float64) *float64 
}
```

 - "ReduceMapper" I think is more clear for its use case
 - There is already a "Result" type and the "Value" property in the code :-/, so I think intput / output is fine
 - I think both funcs can have the same signature? That way when fulfilling the interface people don't need to write the "NewSeries", "AppendPoint" loop etc., just pass the function that is handled in the body of the loop. Then that loop can move to the `Reduce` method.

Reply:
Good idea 👍🏻